[ZIP 307] Light Client Protocol for Payment Detection (#226)
zcash/zips: zip-XXX-light-payment-detection.rst

::

>   ZIP: XXX

Reviewer: ZIP: XXX -> ZIP: 307

Reviewer 2: ```suggestion
  ZIP: 307
```

  Title: Light Client Protocol for Payment Detection
  Authors: George Tankersley <gtank@z.cash>
           Jack Grigg <jack@z.cash>
  Credits: Matthew Green <mgreen@z.cash>
  Category: Standards Track
  Created: 2018-09-17
  License: MIT

Terminology
===========

The key words "MUST", "SHOULD", and "MAY" in this document are to be interpreted as
described in RFC 2119. [#RFC2119]_

The terms below are to be interpreted as follows:

Light client
  A client that is not a full participant in the network of Zcash peers. It can send and
  receive payments, but does not store or validate a copy of the blockchain.

Abstract
========

This proposal defines a protocol for a Zcash light client supporting Sapling shielded
transactions.

Motivation
==========

Currently a client that wishes to send or receive shielded payments must be a full node

> participanting in the Zcash network. This requires an amount of available bandwidth,

Reviewer: Typo: "participanting"

client protocol addresses that need, and is appropriate for low-power,
bandwidth-conscious, or otherwise limited machines (such as mobile phones).

High-Level Design
=================

There are three logical components to a Zcash light client system:

- **Zcash node** that provides chain state and serves as a root of trust for the system.

- **Proxy server** that extracts blockchain data from zcashd to store and serve it in a
  lower-bandwidth format.

- **Light client** that subscribes to the stream from a proxy server and uses that data to
  update its own view of the chain state. The light client MAY be attached to a wallet
  backend that will track particular Sapling notes.

.. figure:: arch.png
    :align: center
    :figclass: align-center

    Outline of the light wallet architecture

Security Model
==============

In this model, we propose **payment detection privacy** as our main security goal. That
is, the proxy should not learn which transactions (received from the blockchain) are
addressed to a given light wallet. If we further assume network privacy (via Tor or
similar), the proxy should not be able to link different connections or queries as

> deriving from the the same wallet.

Reviewer: Typo: "the the"

In particular, the underlying Zcash node / proxy combination is assumed to be "honest but
curious" and is trusted to provide a correct view of the current best chain state and to
faithfully transmit queries and responses. Methods for weakening this assumption are
discussed in appendix FOO.

This ZIP does not address how to spend notes privately.

Compact Stream Format
=====================

A key observation in this protocol is that the current zcashd encrypted field is several
hundred bytes long, due to the inclusion of a transaction “memo”. The need to download

> this entire field imposes a substantial bandwidth cost on each light wallets, which may be

Reviewer: Typo: "on each light wallets"

Reviewer 2: ```suggestion
this entire field imposes a substantial bandwidth cost on each light wallet instance, which may be
```

a limited mobile device on a restricted-bandwidth plan. While more efficient techniques
can be developed in the future, for the moment we propose ignoring the memo field during

> payment detection. Futhermore, we can also ignore any information that is not directly

Reviewer: Typo: "Futhermore"

    // transaction: either a block and an index, or a direct transaction hash.
    message TxFilter {
        BlockID blockID = 1;
        uint64 txIndex = 2;
        bytes txHash = 3;
    }

Client operation
================

Light clients obtain compact blocks from one or more proxy servers, which they then
process locally to update their view of the block chain. We consider only a single proxy
server here without loss of generality.

Local processing
----------------

Given a ``CompactBlock`` received in height-sequential order from a proxy server, a light
client can process it in four ways:

Scanning for relevant transactions
``````````````````````````````````
For every ``CompactOutput`` in the ``CompactBlock``, the light client can trial-decrypt it
against a set of Sapling incoming viewing keys. The procedure for trial-decrypting a
``CompactOutput`` (*cmu*, *epk*, *ciphertext*) with an incoming viewing key *ivk* is a
slight deviation from the standard decryption process [#sapling-ivk-decryption]_:

- let sharedSecret = KA\ :sup:`Sapling`\ .Agree(*ivk*, *epk*)
- let K\ :sup:`enc` = KDF\ :sup:`Sapling`\ (sharedSecret, *epk*)

> - let P\ :sup:`enc` = ChaCha20.Decrypt\ :sub:`K^enc`\ (*ciphertext*)

Reviewer 2: This is not specified precisely enough. The intent is to perform the decryption as it would be performed by IETF_AEAD_CHACHA20_POLY1305, but this involves decrypting with the block counter starting from 1. Since ChaCha20 is not defined at all, it's not clear that the block counter should start with 1.

I suggest adding a definition like this:

```
Let ChaCha20.Decrypt\ :sub:`K^enc`\ (*initialCounter*, *ciphertext*) be the
decryption of *ciphertext* with key K^enc, initial counter value *initialCounter*,
and all-zero nonce, as specified in section 2.4 of [#RFC7539]_.
```
and then use ChaCha20.Decrypt<sub>K^enc</sub>(1, *ciphertext*).

-------------------

The synchronisation phases give the light client sufficient information to determine
accurate address balances, show when funds were received or spent, and spend any unspent
notes. As synchronisation happens via a broadcast medium, it leaks no information about
which transactions the light client is interested in.

If, however, the light client needs access to other components of a transaction (such as
the memo fields for received notes, or the outgoing ciphertexts in order to recover spend
information when importing a wallet seed), it will need to download the full transaction.
The light client SHOULD obscure the exact transactions of interest by downloading numerous
uninteresting transactions as well, and SHOULD download all transactions in any block from
which a single full transaction is fetched (interesting or otherwise). It MUST convey to
the user that fetching full transactions will reduce their privacy.

Appendix FOO
============

You can require the proxy to give you all the block headers to validate.

Reference Implementation
========================

This proposal is supported by a set of libraries and reference code made available by the

> Zcash Company.

Reviewer: Zcash Company -> Electric Coin Company

Reviewer 2: ```suggestion
Electric Coin Company.
```
